src/amsterdam_schema/publish/cli.py

```diff
@@ -86,15 +86,43 @@ def fetch_local_as_publishable(
     return publishable_paths
 
 
-def get_index_file_obj(publishable_paths: List[List[str]]) -> BytesIO:
-    index = {}
-    for path_parts in publishable_paths:
+def get_index_file_obj(publishable_paths: List[List[str]], files_root: Path) -> BytesIO:
+    """Create index file of all the datasets in JSON format.
+
+    For every dataset.json file it will add an entry of the form:
+    `dataset_id:'path/to/dataset'`.
+    When a versions.json file is encountered, its path will be included
+    and any dataset.json files with the same id in subdirectories ignored.
+    """
+    index: Dict[str, str] = {}
+    versions_file = "versions.json"
+    dataset_file = "dataset.json"
+    for path_parts in sorted(publishable_paths, key=len):
         if path_parts[1] != "datasets":
             continue
-        dataset_ext = path_parts[-1]
+        if path_parts[-1] not in (versions_file, dataset_file):
+            continue
+
+        # Grab id from dataset file
+        with open(str(files_root / "/".join(path_parts))) as f:
```

Comment:
No need to call `str` on `Path` object in call to `open`. `open` accepts `Path` objects just fine. Alternatively use [Path.open](https://docs.python.org/3.8/library/pathlib.html?highlight=path%20open#pathlib.Path.open)

Reply:
You are also mixing `Path` operations together with `str` operations. This expression, however, can be fully expressed using `Path` operations.

```python
with files_root.joinpath(Path(path_parts)).open() as f:
   ...
```
or 
```python
with open(files_root.joinpath(Path(path_parts))) as f:
   ...
```

Reply:
Except `path_parts` is a list, so you can't `Path` it. You can do

```python
open(files_root / os.path.join(*path_parts))
```

to get rid of the literal `"/"`.

Reply:
good catch sloppy me.

Reply:
If `path_parts` is a list a destructing operation suffices (forgot that in my previous suggestion):

```python
with open(files_root.joinpath(Path(*path_parts))) as f:
   ...
```
Reason `Path(*path_parts)` works is that Path takes any number of _pathsegments_: https://docs.python.org/3.8/library/pathlib.html?highlight=path#pathlib.PurePath

Reply:
Oh, right, should have checked that.

Reply:
Re: using `os.path.join` to get rid of the literal `"/"`.

Still think we should be using `pathlib` datastructures and operations exclusively and not mix them with string based ones. Especially as `with open(files_root.joinpath(Path(*path_parts))) as f:` works, now that I have added the list destructuring operation.

BTW if you don't like the `joinpath`, `with open(files_root / Path(*path_parts)) as f:` should work equally well. `joinpath` is there specifically to enable a fluent API. Eg. to be able to call `.open()` on it if you want to. But as we are using the built in `open` function, we can suffice with `/` instead of `joinpath`